adding mixed_dtype_mm to torch._inductor

HDCharles opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #106516
* __->__ #106443

Summary: if torch._inductor.config.use_mixed_mm then we can convert
torch.mm(a, b.to(some_dtype)) into a triton kernel where the casting b
is fused into the matmul rather than needing to instantiate the casted b
tensor. If use_mixed_mm is set, this fused kernel will be autotuned
against the default 2 kernel fallback option. If force_mixed_mm then the
fused kernel will always be used, This option is needed for weight-only quantization where we are in
some cases relying on the superior memory characteristics of the fused
kernel rather than the perf numbers (when we can't afford to load memory
 with a tensor 4x the size of our quantized one).

Test Plan: python test/inductor/test_pattern_matcher.py -k "mixed_mm"

python test/inductor/test_torchinductor.py -k "mixed_mm"

Reviewers:

Subscribers:

Tasks:

Tags:

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @Xia-Weiwen @wenzhe-nrv @jiayisunx @peterbell10 @ipiszy @ngimel @yf225 @chenyang78 @kadeng @muchulee8 @aakhundov

HDCharles committed:
adding mixed_dtype_mm to torch._inductor

Summary: if torch._inductor.config.use_mixed_mm then we can convert
torch.mm(a, b.to(some_dtype)) into a triton kernel where the casting b
is fused into the matmul rather than needing to instantiate the casted b
tensor. This is necessary for weight-only quantization where we don't
want to load memory with a tensor 4x the size of our quantized one.

Test Plan: python test/inductor/test_pattern_matcher.py -k "mixed_mm"

python test/inductor/test_torchinductor.py -k "mixed_mm"

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

HDCharles committed:
Update on "adding mixed_dtype_mm to torch._inductor"


Summary: if torch._inductor.config.use_mixed_mm then we can convert
torch.mm(a, b.to(some_dtype)) into a triton kernel where the casting b
is fused into the matmul rather than needing to instantiate the casted b
tensor. This is necessary for weight-only quantization where we don't
want to load memory with a tensor 4x the size of our quantized one.

Test Plan: python test/inductor/test_pattern_matcher.py -k "mixed_mm"

python test/inductor/test_torchinductor.py -k "mixed_mm"

Reviewers:

Subscribers:

Tasks:

Tags:

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 ipiszy ngimel yf225 chenyang78 kadeng muchulee8 aakhundov

[ghstack-poisoned]

HDCharles committed:
Update on "adding mixed_dtype_mm to torch._inductor"


Summary: if torch._inductor.config.use_mixed_mm then we can convert
torch.mm(a, b.to(some_dtype)) into a triton kernel where the casting b
is fused into the matmul rather than needing to instantiate the casted b
tensor. This is necessary for weight-only quantization where we don't
want to load memory with a tensor 4x the size of our quantized one.

Test Plan: python test/inductor/test_pattern_matcher.py -k "mixed_mm"

python test/inductor/test_torchinductor.py -k "mixed_mm"

Reviewers:

Subscribers:

Tasks:

Tags:

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 ipiszy ngimel yf225 chenyang78 kadeng muchulee8 aakhundov

[ghstack-poisoned]

HDCharles committed:
Update on "adding mixed_dtype_mm to torch._inductor"


Summary: if torch._inductor.config.use_mixed_mm then we can convert
torch.mm(a, b.to(some_dtype)) into a triton kernel where the casting b
is fused into the matmul rather than needing to instantiate the casted b
tensor. This is necessary for weight-only quantization where we don't
want to load memory with a tensor 4x the size of our quantized one.

Test Plan: python test/inductor/test_pattern_matcher.py -k "mixed_mm"

python test/inductor/test_torchinductor.py -k "mixed_mm"

Reviewers:

Subscribers:

Tasks:

Tags:

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 ipiszy ngimel yf225 chenyang78 kadeng muchulee8 aakhundov

[ghstack-poisoned]

HDCharles commented:
> Also curious what error is thrown when a user does a bad cast?
> a is fp32 and b is int8 and you cast b to fp16.

it'll throw an error that the dtypes need to match for tl.dot op, unless there are some combinations that dot can work with, not sure if tl.dot has support for something weird like fp32, fp64 or something like that

ezyang commented:
This seems... fine? But I'm not the main owner for the code here.

jansel commented on torch/_inductor/kernel/mm.py:
```diff
@@ -199,3 +201,18 @@ def tuned_addmm(inp, mat1, mat2, *, alpha=1, beta=1, layout=None):
+def tuned_mixed_mm(mat1, mat2, mat2_dtype):
+    m, n, k, layout, mat1, mat2 = mm_args(mat1, mat2, layout=None)
+    # options to tune from
+    choices = []
```
Should we have a choice to fallback to aten (with two kernels)?

HDCharles replied:
how would i do that? atm its gated by config only

jansel commented on torch/_inductor/kernel/mm.py:
```diff
@@ -199,3 +201,18 @@ def tuned_addmm(inp, mat1, mat2, *, alpha=1, beta=1, layout=None):
+    choices = []
+    b_prologue_cast_type = f"tl.{mat2_dtype}".replace("torch.", "")
+    for config in mm_configs(m, n, k):
+        mm_template.maybe_append_choice(
```
Will this work on CPU?

HDCharles replied:
no, its similar to tuned_int_mm

HDCharles replied:
do you think it should be gated so it doesn't match on CPU?

jansel replied:
I think it would be fine to just fall back to the eager ops on CPU, but it should at least run without error and be tested.

HDCharles replied:
ok done

jansel commented on test/inductor/test_torchinductor.py:
```diff
@@ -1745,6 +1745,38 @@ def fn(a, b):
             check_lowp=False,
         )
 
+    @config.patch(use_mixed_mm=True)
+    def test_mixed_mm(self):
+        def fn(a, b):
+            return torch.mm(a, b.to(a.dtype))
+
+        if self.device == "cuda":
+            self.common(
+                fn,
+                (
+                    torch.randn(8, 8),
+                    torch.randint(-128, 127, (8, 8), dtype=torch.int8),
+                ),
+                check_lowp=True,
+            )
+
+    @config.patch(use_mixed_mm=True)
+    def test_mixed_mm2(self):
+        def fn(a, b, scale, bias):
+            return torch.mm(a, b.to(a.dtype)) * scale + bias
+
+        if self.device == "cuda":
+            self.common(
+                fn,
+                (
+                    torch.randn(8, 8),
+                    torch.randint(-128, 127, (8, 8), dtype=torch.int8),
+                    torch.randn(8),
+                    torch.randn(8),
+                ),
+                check_lowp=True,
+            )
+
```
We should include a test that the kernel/code is actually being used.  See test_select_algorithm for examples.

HDCharles replied:
there is a test in the pattern matcher that checks it, it seems like it checks the same thing, right?

jansel commented on test/inductor/test_torchinductor.py:
```diff
@@ -1745,6 +1745,38 @@ def fn(a, b):
+        def fn(a, b):
+            return torch.mm(a, b.to(a.dtype))
+
+        if self.device == "cuda":
```
We should test CPU

HDCharles replied:
ok, i'll add that, but i think it just errors iirc

jansel reviewed (changes requested):
Some comments inline.  Will these patterns also fire for AMP?  If so can we do a perf run?

HDCharles commented:
> Some comments inline. Will these patterns also fire for AMP? If so can we do a perf run?

I'm not sure if it will hit amp in general, but it certainly won't hit unless the user enables it in the config (disabled by default) is it still needed in that case?

HDCharles committed:
Update on "adding mixed_dtype_mm to torch._inductor"


Summary: if torch._inductor.config.use_mixed_mm then we can convert
torch.mm(a, b.to(some_dtype)) into a triton kernel where the casting b
is fused into the matmul rather than needing to instantiate the casted b
tensor. This is necessary for weight-only quantization where we don't
want to load memory with a tensor 4x the size of our quantized one.

Test Plan: python test/inductor/test_pattern_matcher.py -k "mixed_mm"

python test/inductor/test_torchinductor.py -k "mixed_mm"

Reviewers:

Subscribers:

Tasks:

Tags:

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 ipiszy ngimel yf225 chenyang78 kadeng muchulee8 aakhundov

[ghstack-poisoned]

HDCharles committed:
Update on "adding mixed_dtype_mm to torch._inductor"


Summary: if torch._inductor.config.use_mixed_mm then we can convert
torch.mm(a, b.to(some_dtype)) into a triton kernel where the casting b
is fused into the matmul rather than needing to instantiate the casted b
tensor. This is necessary for weight-only quantization where we don't
want to load memory with a tensor 4x the size of our quantized one.

Test Plan: python test/inductor/test_pattern_matcher.py -k "mixed_mm"

python test/inductor/test_torchinductor.py -k "mixed_mm"

Reviewers:

Subscribers:

Tasks:

Tags:

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 ipiszy ngimel yf225 chenyang78 kadeng muchulee8 aakhundov

[ghstack-poisoned]

HDCharles committed:
Update on "adding mixed_dtype_mm to torch._inductor"


Summary: if torch._inductor.config.use_mixed_mm then we can convert
torch.mm(a, b.to(some_dtype)) into a triton kernel where the casting b
is fused into the matmul rather than needing to instantiate the casted b
tensor. This is necessary for weight-only quantization where we don't
want to load memory with a tensor 4x the size of our quantized one.

Test Plan: python test/inductor/test_pattern_matcher.py -k "mixed_mm"

python test/inductor/test_torchinductor.py -k "mixed_mm"

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jansel commented on torch/_inductor/fx_passes/post_grad.py:
```diff
@@ -152,6 +152,29 @@ def mm_plus_mm(match: Match, mat1, mat2, mat3, mat4):
+        ),
+    ),
+    extra_check=(
+        lambda match: config.use_mixed_mm and match.kwargs["mat1"].meta["val"].is_cuda
```
```suggestion
        lambda match: config.use_mixed_mm and getattr(match.kwargs["mat1"].meta.get("val"), "is_cuda", False)
```

HDCharles replied:
fixed

HDCharles committed:
Update on "adding mixed_dtype_mm to torch._inductor"


Summary: if torch._inductor.config.use_mixed_mm then we can convert
torch.mm(a, b.to(some_dtype)) into a triton kernel where the casting b
is fused into the matmul rather than needing to instantiate the casted b
tensor. If use_mixed_mm is set, this fused kernel will be autotuned
against the default 2 kernel fallback option. If force_mixed_mm then the
fused kernel will always be used, This option is needed for weight-only quantization where we are in
some cases relying on the superior memory characteristics of the fused
kernel rather than the perf numbers (when we can't afford to load memory
 with a tensor 4x the size of our quantized one).

Test Plan: python test/inductor/test_pattern_matcher.py -k "mixed_mm"

python test/inductor/test_torchinductor.py -k "mixed_mm"

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

HDCharles committed:
Update on "adding mixed_dtype_mm to torch._inductor"


Summary: if torch._inductor.config.use_mixed_mm then we can convert
torch.mm(a, b.to(some_dtype)) into a triton kernel where the casting b
is fused into the matmul rather than needing to instantiate the casted b
tensor. If use_mixed_mm is set, this fused kernel will be autotuned
against the default 2 kernel fallback option. If force_mixed_mm then the
fused kernel will always be used, This option is needed for weight-only quantization where we are in
some cases relying on the superior memory characteristics of the fused
kernel rather than the perf numbers (when we can't afford to load memory
 with a tensor 4x the size of our quantized one).

Test Plan: python test/inductor/test_pattern_matcher.py -k "mixed_mm"

python test/inductor/test_torchinductor.py -k "mixed_mm"

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

HDCharles commented:
@pytorchmergebot merge

HDCharles commented:
@pytorchmergebot merge
